### PR TITLE
feat: Context Engineering Layer (view.py) + Unity Bug Fixes

### PR DESCRIPTION
**Pull request by vanillaSky00**
## Summary

Two tightly coupled changesets that must land together:

1. **Backend** — Introduces `view.py` as a new context engineering layer. The raw Unity perception payload that agents previously consumed directly is now transformed into a structured, affordance-driven 7-block perception string. All rule-based logic previously embedded as static prose in prompt templates is now data-driven and produced from live game state.

2. **Unity** — Fixes `put_down` rejections on the assembly plate, silent empty-pickup from `TomatoBox`, and incorrect scene serialization on `Plate.cs`. Shifts ownership of burger-assembly state from Python into Unity, making the backend a thin consumer.

> These two changesets are paired because the Unity-side assembly state ownership transfer directly supersedes the Python-side `assembly.py` tracker — landing them together keeps `main` consistent.

---

## Architecture: Before → After

```
BEFORE
  Unity → routes.py → agents (parse raw fields directly)
                    → prompt templates (static prose rules embedded)

AFTER
  Unity → view.py (KitchenRegistry + affordance rendering → 7-block perception string)
        → routes.py (build_perception_context once per frame, thread via AgentState)
        → all agents (read structured [LAYOUT] / [A]–[F] blocks)
        → prompt templates (format guidance only — no situational rules)
```

---

#### `KitchenRegistry` — Python-owned domain knowledge

Compensates for Unity's inconsistent naming across fields:

- `RAW_INGREDIENTS`, `PROCESSED_INGREDIENTS` — canonical item sets
- `PROCESSING_RULES` — maps each raw ingredient to its required station and action verb
- `CONTAINER_BOXES`, `STATIONS` — exact object ID sets Unity expects
- `SHARED_PREP_TABLES`, `PLAYER_PREP_TABLES` — correct canonical IDs (`Preparation1`..`4`, `Player_preparation_1`..`2`), fixing the `Preparation_1` underscore hallucination baked into old templates
- `HAMBURGER_STACK` — authoritative placement-order tuple
- `is_plate` / `is_processed` / `is_raw` helpers — tolerant of Unity name variants (`COOKEDMEAT`, `SLICED_BREAD`, `SLICEDBREAD`, `SLICEDON`)

#### `HeldItem` dataclass

Typed, frozen representation of what the agent is currently holding. Decouples all downstream code from `self.held_item`'s unstable schema (bare string, nested object, or null). Exposes `is_raw`, `is_processed`, `is_empty_hands` as properties.

#### `ObservationAdapter`

Thin accessor layer over the `Perception` Pydantic model. Provides stable property contracts (`location`, `held_item`, `reachable`, `visible`, etc.) so renderers and agents never touch raw schema fields directly.

#### `PerceptionRenderer`

The core context engineering class. Produces a 7-component perception block injected into all three active agents each frame:

| Block | Content | What it replaces in old templates |
|---|---|---|
| `[LAYOUT]` | Canonical object IDs + burger assembly flow | Hardcoded `--- UNITY OBJECTS ---` lists; wrong `Preparation_1` IDs |
| `[A]` Self state | Held item + hands-full status | `ONE ITEM LIMIT / HANDS-FULL PROTOCOL` static rule |
| `[B]` Immediate affordances | Per-object action guidance from live state | `PHYSICS RULES: if on CutBoard, must chop`; `PROCESSING RULES` block |
| `[C]` Kitchen state | Supply check: ready / raw / plated per table | `SUPPLY CHECK: do not re-gather`; `ALLOW LOGICAL STATE CHANGES` |
| `[D]` Short-term memory | Last 4 execution trace steps | Per-agent history prose |
| `[E]` Failure context | Retry count + hint + hard-stop warning | `RETRIES COUNT rule` |
| `[F]` Assembly progress | Live plate location / stack / next expected from `perception.assembly` | `HAMBURGER stacking order` static text |

Each `render_*` method turns a rule into an unavoidable physical reality in context ("The CutBoard holds a raw TOMATO. You MUST chop it immediately.") rather than an abstract instruction competing with everything else in the prompt.

#### `build_perception_context` — module-level shortcut

```python
def build_perception_context(
    perception: Perception,
    retry_count: int = 0,
    current_task: str = "",
) -> str:
    return PerceptionRenderer(perception).build_perception_context(
        retry_count=retry_count,
        current_task=current_task,
    )
```

Called once per frame in `routes.py` and threaded through `AgentState`. All agents share the same block — they cannot disagree about kitchen state.

---

### `routes.py`
- Imports `build_perception_context` from `view.py`
- Calls it once per frame, stores result on `AgentState.context`
- Agents read `state.context` instead of importing from `app.context.view` directly
- Logger cleanup: correct levels (`info` vs `warning`), lazy `%`-formatting, `logger.exception` on fatal paths, fixed `broadcast()` missing `await`

### `schemas.py`
- Added `AssemblyView` Pydantic model (`plate_location`, `stack`, `next_expected`, `is_done`)
- Added `Perception.assembly: AssemblyView` field to receive Unity's authoritative assembly state

### `exceptions.py`
- Typed hierarchy: `PaprikaError` base → `InvalidPerceptionError` / `ContextBuildError` / `AgentExecutionError`
- Single catch-translate-respond block in the websocket loop

### `action.md`, `critic.md`, `curriculum.md`
- Removed static `PHYSICS RULES`, `HAMBURGER stacking order`, `ONE ITEM LIMIT`, `SUPPLY CHECK`, `RETRIES COUNT`, `Preparation_1..4` blocks
- Templates now describe response *format* only — all situational rules come from the live `[A]`–`[F]` perception block
- `"Explore the area"` curriculum fallback replaced with `"Set up the assembly plate on Preparation1"` — concrete and verifiable

---

## Backend — Bug Fixes

**`Preparation_1` hallucination** — Templates taught the wrong ID. `[LAYOUT]` now emits `Preparation1` (no underscore) and all example JSON uses the correct form.

**Raw-on-plate confusion** — `[B]` container affordances explicitly label box contents as `"RAW — must be chopped/cooked"`. `critic.md` Rule 3b marks raw-on-plate as automatic fail.

**Supply check reliability** — `render_supply_check` derives primarily from live `held_item` on visible/reachable prep tables; falls back to `statistics.table_items` aggregate only when no prep table is in sight (Unity's aggregate is unreliable per-frame).

**Hands-full retry cascade** — `_correction_hint` distinguishes two failure modes:
- `put_down` with empty hands → a prior `pickup` silently failed; prescribes redundant `move_to → move_to → pickup` pattern
- `pickup` from an empty source → the container ran out; pick a different source

**Retry threshold** — `[E]` surfaces consecutive failure count and emits a hard stop warning at ≥2 retries, replacing the implicit infinite retry loop.

**JSON parser (`base.py`)** — Root cause of curriculum being called 4× per frame and phantom `Explore the area` fallbacks. The list-regex matched inner brackets in `[C]` / `[B]` first, causing `json.loads("[C]")` to raise before the object branch was tried. Fixed: pure-JSON → code-fence strip → balanced-bracket extraction with string-aware depth counting.

**Task reshuffling loop** — Critic now recognises "task already satisfied" when a ready form is on any parking surface, with case-insensitive name matching; `action.md` Rule 4 forces an empty plan when `[C]` already shows the target.

---

## Backend — Deletions Driven by Unity Ownership Transfer

| What | Lines removed |
|---|---|
| `backend/app/context/assembly.py` deleted | ~227 lines (`AssemblyState`, `observe_plate`, `record_task_success`, `_trace_*` helpers, `empty_state_dict`) |
| `routes.py` | Dropped assembly import, `session_state` key, `observe_plate` pre-pass, `record_task_success` post-pass |
| `view.py` | `_render_assembly_progress` now reads `perception.assembly` directly; `build_perception_context` no longer takes an assembly dict |

---

### `Plate.cs` — scene-reset bug
Both plates in `GameScene.unity` shipped with `currentObjectIndex: 5` serialised. Added `Start() { ResetPlate(); }` so every play session starts at slot 0. Exposed `PlacedStack` / `NextExpectedType` for downstream reporting.

### `Inventory.cs` — `TakeItem` silent-fail
`TakeItem` now returns `bool`. `ActionPickup` reads the return value and surfaces `"Hands full — already holding X. Put it down first."` instead of falsely reporting pickup success when hands were already full.

### `ActionPickup.cs` — wrong `ItemBox` subclass
`GetComponent<ItemBox>()` could land on a `Table`/`TableBox` sibling whose overridden `GetItem()` returns `NONE`, making crates appear empty. New `PickIngredientBox()` prefers the plain `ItemBox` base class (on self, then children) before falling back to subclasses. Added a `Debug.Log` reporting which concrete type was picked and what `GetItem()` returned.

### `ActionPut.cs` — messaging
Success path now emits:
```
Stacked BreadSlice on Preparation1 (plate 1/7). Next expected: CookedMeat.
```
Failure path distinguishes plate-assembly (`plate rejected X — next expected is Y`) from non-plate (`already occupied`). Resolves either `Table` or `TableBox` wrapper as the assembly target.

---

## Unity — Assembly State Ownership Transfer

New accessors on `Table.cs` / `TableBox.cs`:

```
IsAssemblySurface / AssemblyStack / NextExpectedAssemblyType / IsAssemblyDone
```

`AgentNetworkManager.cs` builds a top-level `assembly` payload by scanning all Tables each frame:

```json
{
  "plate_location": "Preparation1",
  "stack": ["BreadSlice", "CookedMeat"],
  "next_expected": "TomatoSlice",
  "is_done": false
}
```

State is reported even when the plated table is outside the agent's vision radius. A canonical name map normalises `SLICEDBREAD → BreadSlice`, `COOKEDMEAT → CookedMeat`, etc. — Python no longer re-normalises. `assembly.py` is deleted as a result.

---

### Automated

```bash
make test   # existing websocket tests — tests/api/test_routes.py
```

### Manual Steps

```
Step 1  docker compose up -d --build  →  connect Unity client
        Open GameScene.unity; confirm no C# compile errors in Console

Step 2  Watch one full burger cycle
        [F] should advance: None → PLATE_SETUP → 7 layers
            (BreadSlice → CookedMeat → TomatoSlice → LettuceSlice → OnionSlice → CheeseSlice → BreadSlice)
        Agent bubble shows Online
        Backend [F] block reflects Unity-reported plate_location / stack / next_expected each frame

Step 3  Go to LangSmith to monitor perception blocks and tool call traces
```

Manual verification checklist:
- `[F]` reads `None → PLATE_SETUP → 7 layers` correctly as Unity reports
- `[B]` affordances fire correctly per station (Oven cooked-ready, CutBoard raw-item, container supply check)
- `[A]` correctly reflects held item and hands-full state each frame
- `[E]` retry hint fires and escalates to hard-stop at ≥2 consecutive failures
- No agent emits `Preparation_1` (underscore) in any action JSON

---

## Checklist

- [ ] Python: `uv run ruff check` clean on `view.py`, `routes.py`, `schemas.py`, `exceptions.py`
- [ ] Unity: C# compiles with no errors in Editor Console
- [ ] `make test` passes
- [ ] Manual steps 1–3 above completed
- [ ] No unintended console errors or warnings (diagnostic `Debug.Log` in `ActionPickup` is intentional)
- [ ] Branch is `feat/context-management-and-fix` — not pushed directly to `main`
- [ ] No new pytest files yet — assembly tracker coverage worth adding as follow-up

### CODE REVIEW

**Comment by vanillaSky00** [comment]
This PR is still in revise, the before after is not right. Plus, the future work is that once the agent done the hambergur do not make it clean, rather we need handle to the customer, sell area. But this is not done and consider yet.

[Edit]: I have revised
